feat: Add current_time Presto function

anandamideShakyan opened the pull request:
In the past, we were not implementing functions like current_time that returned constants in Velox. This was because Presto Java would infer the constant value at the co-ordinator and fold it, so the worker was never passed this function in execution.

However, with recent changes for side-car and worker based constant folding implementation, we need native implementations for functions like current_time, so adding to Velox as well.

aditi-pandit commented on velox/docs/functions/presto/datetime.rst:
```diff
@@ -188,6 +188,10 @@ Date and Time Functions
 
         SELECT current_timezone;   -- Asia/Kolkata
 
+.. function:: current_time() -> time with time zone
```
This function is alphabetically before current_timezone. Please move it ahead of it.

anandamideShakyan replied:
Done.

aditi-pandit commented on velox/functions/prestosql/tests/DateTimeFunctionsTest.cpp:
```diff
@@ -6833,3 +6833,44 @@ TEST_F(DateTimeFunctionsTest, currentTimezone) {
+  auto laPacked = callCurrentTime(1710061200000, "America/Los_Angeles");
+  // Reuse TimestampWithTimezone: TIME WITH TIME ZONE uses the same
+  // packed (millis, zoneKey) format, so unpack() works.
+  auto la = TimestampWithTimezone::unpack(laPacked);
```
Can you add bunch more test cases in different timezones ?

anandamideShakyan replied:
Done.

aditi-pandit reviewed (commented):
Thanks @anandamideShakyan. Have few comments.

anandamideShakyan commented on velox/functions/prestosql/DateTimeFunctions.h:
```diff
@@ -2031,4 +2032,29 @@ struct LocalTimeFunction {
+    result = result_;
+  }
+private:
+  int64_t currentTimeSinceMidnight_;
```
You are right, it is not needed will remove it. I was following the implementation of local_time since current_time is just local_time + timezone. I also had a doubt about handling offset and legacytimestamp like it is done in presto java. The velox implementation of local_time doesn't handle legacytimestamp case also it doesn't adjust the offset.  Why do the implementations diverge, are they handled differently in velox , am I missing something here ?

czentgr replied:
The code from the comment is code is from the Java `localtime` and not `current_time` implementation. The `current_time` implementation is here: https://github.com/prestodb/presto/blob/327032c0e8a01d6ce554847036d81a853a562e11/presto-main-base/src/main/java/com/facebook/presto/operator/scalar/DateTimeFunctions.java#L114
It uses UTC and the timezone.
The `localtime` function would convert the session start time (in UTC) to local time using the time zone information.
Also `legacyTimestamp` doesn't come into the picture for `current_time`.

anandamideShakyan commented on velox/functions/prestosql/DateTimeFunctions.h:
```diff
@@ -2031,4 +2032,29 @@ struct LocalTimeFunction {
+struct CurrentTimeFunction {
+  VELOX_DEFINE_FUNCTION_TYPES(T);
+  int64_t result_{0};
+  const tz::TimeZone* timeZone_ = nullptr;
```
Changed to local variable.

czentgr commented on velox/functions/prestosql/tests/DateTimeFunctionsTest.cpp:
```diff
@@ -6816,6 +6816,62 @@ TEST_F(DateTimeFunctionsTest, dateAddDateVariableUnit) {
+    auto rowVector = makeRowVector({});
+    rowVector->resize(1);
+
+    auto result = evaluate("current_time()", rowVector);
```
You should use `evaluateOnce<int64_t>` instead. That reduces the code quite a bit using this helper.

czentgr reviewed (commented):
Please also run the fuzzer with the new function and post the results. Instructions are here (step 7): https://github.com/facebookincubator/velox/blob/main/CONTRIBUTING.md#adding-scalar-functions

Note, the CI does show fuzzer failures. please take a look.
The function is non-deterministic and needs to be excluded from the expression fuzzer?

aditi-pandit commented on velox/functions/prestosql/registration/DateTimeFunctionsRegistration.cpp:
```diff
@@ -346,6 +346,8 @@ void registerSimpleFunctions(const std::string& prefix) {
       {prefix + "parse_duration"});
 
   registerFunction<LocalTimeFunction, Time>({prefix + "localtime"});
+  registerFunction<CurrentTimeFunction, TimeWithTimezone>(
```
Is this an alias for now ? If yes, then we should add the alias as well.

anandamideShakyan commented on velox/expression/fuzzer/ExpressionFuzzerTest.cpp:
```diff
@@ -450,7 +450,8 @@ std::unordered_set<std::string> skipFunctionsSOT = {
     "$internal$contains",
-    "localtime", // localtime cannot be called with paranthesis:
+    "current_time", // current_time cannot be called with parenthesis
+    "localtime", // localtime cannot be called with parenthesis:
```
fixed typo: parenthesis

kgpai commented on velox/functions/prestosql/DateTimeFunctions.h:
```diff
@@ -2113,4 +2113,44 @@ struct LocalTimeFunction {
+    int64_t currentTimeSinceMidnight = sessionStartTimeMs % kMillisInDay;
+    const tz::TimeZone* timeZone = getTimeZoneFromConfig(config);
+    if (timeZone == nullptr) {
+      VELOX_USER_FAIL("Timezone cannot be null");
```
nit: Can you validate that java doesnt allow the timezone to not be set ?

anandamideShakyan replied:
Validated Java behavior — Presto session always has a TimeZoneKey (SqlFunctionProperties.getTimeZoneKey() is required). Updated to VELOX_CHECK and added comment.

anandamideShakyan commented on velox/functions/prestosql/DateTimeFunctions.h:
```diff
@@ -2113,4 +2113,44 @@ struct LocalTimeFunction {
+      utcMillisOfDay += kMillisInDay;
+    }
+
+    int32_t offsetAtInstant =
```
In the java implementation we take the utcmillis and adjust it by difference of : ( dst offset at epoch and at offset at sessionStartMs), I have followed the same in my implementation. 
https://github.com/prestodb/presto/blob/873393fb0123fb8d0621da9fd877b19faf53b831/presto-main-base/src/main/java/com/facebook/presto/operator/scalar/DateTimeFunctions.java#L118C9-L125C1

I have just done this `valueToSessionTimeZoneOffsetDiff()`  calculation below. 

Do you want me to convert to local wall time and store the offset between local and utc time as timezone key ? That would also be accurate because both will give the same values:
Whether we adjust utcmillisofday for offset difference and store the timezonekey, or we store the localmillis along with difference of utcmillis and localmillis(basically current time difference). I think this was done by Natasha in her PR. 

I still think my implementation follows the java implementation more closely, if you still want me to change it let me know.

kgpai commented:
@anandamideShakyan Will do.

anandamideShakyan commented:
@kgpai I have addressed the comments. About std::chrono:: readability issue, I see that my implementation is consistent with velox coding style so I haven't made any changes( many other functions in the same file have implemented it without explicitly adding a namespace qualifier). Let me know what you think. PTAL again. Thanks. 

Also sorry if I’ve been nudging a bit too often for reviews — I appreciate your time and feedback.
